mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassin.java

```diff
@@ -30,32 +31,76 @@
+    private Optional<UnicastProcessor<Tuple2<InputStream, Username>>> spamQueue;
+    private Optional<Disposable> hamHandle;
+    private Optional<Disposable> spamHandle;
 
     @Inject
     public SpamAssassin(MetricFactory metricFactory, SpamAssassinConfiguration spamAssassinConfiguration) {
         this.metricFactory = metricFactory;
-        this.spamAssassinConfiguration = spamAssassinConfiguration;
+
+        createListeners(spamAssassinConfiguration);
     }
 
-    public void learnSpam(List<InputStream> messages, Username username) {
-        if (spamAssassinConfiguration.isEnable()) {
-            Host host = spamAssassinConfiguration.getHost().get();
-            SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
-            messages
-                .forEach(Throwing.consumer(message -> invoker.learnAsSpam(message, username)));
-        }
+    public void learnSpam(Flux<InputStream> messages, Username username) {
+        spamQueue.ifPresent(queue ->
+            messages.zipWith(Mono.just(username).repeat())
+                .doOnNext(queue::onNext)
+                .then()
+                .block()
+        );
     }
 
-    public void learnHam(List<InputStream> messages, Username username) {
+    public void learnHam(Flux<InputStream> messages, Username username) {
+        hamQueue.ifPresent(queue ->
+            messages.zipWith(Mono.just(username).repeat())
+                .doOnNext(queue::onNext)
+                .then()
+                .block()
+        );
+    }
+
+    @Override
+    public void close() {
+        hamHandle.ifPresent(Disposable::dispose);
+        spamHandle.ifPresent(Disposable::dispose);
+    }
+
+    private void createListeners(SpamAssassinConfiguration spamAssassinConfiguration) {
         if (spamAssassinConfiguration.isEnable()) {
             Host host = spamAssassinConfiguration.getHost().get();
             SpamAssassinInvoker invoker = new SpamAssassinInvoker(metricFactory, host.getHostName(), host.getPort());
-            messages
-                .forEach(Throwing.consumer(message -> invoker.learnAsHam(message, username)));
+
+            UnicastProcessor<Tuple2<InputStream, Username>> ham = UnicastProcessor.create();
+            hamQueue = Optional.of(ham);
+            hamHandle = Optional.of(ham
+                .doOnNext(Throwing.consumer(messageUsername -> invoker.learnAsHam(messageUsername.getT1(), messageUsername.getT2())))
+                .subscribeOn(Schedulers.elastic())
+                .subscribe());
+
+            UnicastProcessor<Tuple2<InputStream, Username>> spam = UnicastProcessor.create();
+            spamQueue = Optional.of(spam);
+            spamHandle = Optional.of(spam
+                .doOnNext(Throwing.consumer(messageUsername -> invoker.learnAsHam(messageUsername.getT1(), messageUsername.getT2())))
+                .subscribeOn(Schedulers.elastic())
+                .subscribe());
```

Comment:
I'd rather have subscribe calls in an init method of a startable object.

Reply:
I tend to avoid  statefull objects

Reply:
but it's how James work! We first wire everything, then start services in the right order.
With this code, you can start to call services that are not started yet (and then fail)

Reply:
I get it, but it's not the case here.